Jasmine/Tetris/ViewModels/TetrisGameViewModel.swift

```diff
@@ -1,4 +1,159 @@
+                tetrisGrid.add(at: newCoordinate, tileText: text)
+                coordinatesToShift.append((from: currentCoordinate, to: newCoordinate))
+            }
+        }
+        delegate?.animate(shiftTiles: coordinatesToShift)
+    }
+
+    // TODO: fetch from database to match valid phrase
+    private func getValidPhraseRange(_ line: String) -> CountableRange<Int>? {
+        let phrases = ["先发制人"]
+        for phrase in phrases {
+            let phraseLen = phrase.characters.count
+            for i in 0...line.characters.count - phraseLen {
+                let startIndex = line.index(line.startIndex, offsetBy: i)
+                if line[startIndex..<line.index(startIndex, offsetBy: phraseLen)] == phrase {
+                    return i..<i + phraseLen
+                }
+            }
+        }
+        return nil
+    }
+
+    // TODO: generate from database, base on existing grid
+    fileprivate func getRandomWord() -> String {
+        let words = "先发制人"
+        return String(words[words.index(words.startIndex,
+                                        offsetBy: Random.integer(toInclusive: UInt(words.characters.count)))])
+    }
+}
+
+extension TetrisGameViewModel: TetrisGameViewModelProtocol {
+
+    func shiftFallingTile(to coordinate: Coordinate) -> Bool {
+        return !tetrisGrid.hasTile(at: coordinate)
+    }
+
+    func dropNextTile() -> (location: Coordinate, tileText: String) {
+        upcomingTiles.append(getRandomWord())
+        let tileText = upcomingTiles.removeFirst()
+        delegate?.redisplayUpcomingTiles()
+        fallingTileText = tileText
+        let randCol = Random.integer(toInclusive: UInt(Constants.Tetris.columns))
+        return (location: Coordinate(row: Coordinate.origin.row, col: randCol),
+                tileText: tileText)
+    }
+
+    func landFallingTile(at coordinate: Coordinate) {
+        guard let fallingTileText = fallingTileText else {
+            assertionFailure("No falling tile")
+            return
+        }
+        tetrisGrid.add(at: coordinate, tileText: fallingTileText)
+
+        guard let destroyedCoordinates = checkForMatchingPhrase() else {
+            return
+        }
+        tetrisGrid.remove(at: destroyedCoordinates)
+        delegate?.animate(destroyTilesAt: destroyedCoordinates)
+
+        currentScore += destroyedCoordinates.count
+        delegate?.redisplay(newScore: currentScore)
+
+        shiftDownTiles(destroyedCoordinates)
+    }
+
+    func swapCurrentTileWithUpcomingTile(at index: Int) {
+        guard let currentFallingTileText = fallingTileText else {
+            assertionFailure("No falling tile")
+            return
+        }
+        fallingTileText = upcomingTiles[index]
+        upcomingTiles[index] = currentFallingTileText
+        delegate?.redisplayUpcomingTiles()
+    }
+
+    func startGame() {
+        countDownTimer.startTimer(timerInterval: Constants.Tetris.timeInterval, viewControllerDelegate: delegate)
```

Comment:
Should guard status too, in case VC decides to call multiple time (of course VC will guard too)